### PR TITLE
handle cross-compilation test failure more gracefully

### LINKED ISSUE #3086: Ideal situation for cross-compilation tests
From [Issue #3084](https://github.com/rust-lang/cargo/issues/3084#issuecomment-246426248), in @alexcrichton's  opinion the ideal situation would be:
- Synchronize all cross-tests to check for a cross-compiler at the beginning.
- If one is found, run all tests
- If one is not found, only fail one test but have all the others pass
- In the failing test, try to detect rustup and recommend rustup target add ....
- If rustup is not found, then print out that this is a failure that can be ignored or you can set an environment variable to ignore the test.

### PR DESCRIPTION
**Pull request by froydnj**
The current cross-compilation test module fails every test if an appropriate cross standard library is not found.  This behavior is unhelpful, because the failures are verbose, unexpected, and there's no
obvious way to make them go away.

Instead, it would be better to check once before all the tests that an appropriate cross-compilation setup is available.  Once that check has been done, a single test can fail with an appropriate `panic!` message while the other tests silently pass.  The `panic!` message can inform the user about their options, either setting an appropriate environment variable, or using rustup to install the necessary cross standard library.

Assuming the user has rustup installed, the single failure now looks
something like:

```
	thread 'plugin_deps' panicked at 'Cannot cross compile to i686-unknown-linux-gnu.

This failure can be safely ignored. If you would prefer to not see this
failure, you can set the environment variable CFG_DISABLE_CROSS_TESTS to "1".

Alternatively, you can install the necessary libraries for cross-compilation with

    rustup toolchain install stable-i686-unknown-linux-gnu
', tests/cross-compile.rs:87
note: Run with `RUST_BACKTRACE=1` for a backtrace.
```

The code is admittedly gnarly, with synchronization and `unsafe` sections, but I wasn't sure how to make it any better. Suggestions welcome!

Fixes #3086.

### CODE REVIEW

File: tests/cross-compile.rs [hunk 1]
```diff
@@ -19,9 +22,79 @@ fn disabled() -> bool {
...
+        // be disabled.
+        return false;
+    }
+
+    // We can't compile a simple cross project.  We want to warn the user
+    // by failing a single test and having the remainder of the cross tests
+    // pass.  We don't use std::sync::Once here because panicing inside its
+    // call_once method would poison the Once instance, which is not what
+    // we want.
+    static mut HAVE_WARNED: AtomicBool = ATOMIC_BOOL_INIT;
```
Comment (alexcrichton) [inline tests/cross-compile.rs]: I believe this doesn't need to be `mut`, if it's just a `static` you can safely access it

File: tests/cross-compile.rs [hunk 2]
```diff
@@ -19,9 +22,79 @@ fn disabled() -> bool {
...
+
+    match unsafe {
+        HAVE_WARNED.compare_exchange(false, true,
```
Comment (alexcrichton) [inline tests/cross-compile.rs]: Perhaps just:

```rust
if HAVE_WARNED.swap(true, SeqCst) {
    return true
}

// do the warning
// ...
```

File: tests/cross-compile.rs [hunk 3]
```diff
@@ -19,9 +22,79 @@ fn disabled() -> bool {
...
+        // We are responsible for warning the user, which we do by panicing.
+        Ok(false) => {
+            let rustup_available = Command::new("rustup").output().is_ok();
+
+            let rustup_help = if rustup_available {
+                format!("
+
+Alternatively, you can install the necessary libraries for cross-compilation with
+
+    rustup toolchain install stable-{}", cross_target)
```
Comment (alexcrichton) [inline tests/cross-compile.rs]: I think this just needs to be `rustup target add $target`, right?

**Comment by alexcrichton** [comment]
Awesome idea! Sounds great to me

**Comment by froydnj** [comment]
Thanks for the review.  Made the suggested changes and added a note for Linux that you might need to install extra libraries, since when I tried the `rustup` advice straight-up, it didn't actually work. 😊 

r? @alexcrichton

**Pull request merged**